File: rnn/elixir/rnn.exs
defmodule RNN do
  use LinearAlgebra

  @doc """
  Creates a new RNN model with the given parameters.

  ## Parameters

  * `input_dim`: The number of features in the input data.
  * `hidden_dim`: The number of neurons in the hidden layer.
  * `output_dim`: The number of features in the output data.

  ## Returns

  A new `RNN` model.
  """
  def new(input_dim, hidden_dim, output_dim) do
    Weights = Vector.fill(input_dim, fn _ -> Vector.fill(hidden_dim, rand()) end)
    Biases = Vector.fill(hidden_dim, rand())
    h = Vector.fill(hidden_dim, 0.0)
    {Weights, Biases, h}
  end

  @doc """
  Trains the RNN model on the given data.

  ## Parameters

  * `data`: The training data.

  ## Returns

  The trained RNN model.
  """
  def train(model, data) do
    for sequence in data do
      for t in sequence do
        z = Weights * t + Biases
        h = tanh(z)
      end
    end
    model
  end

  @doc """
  Makes predictions on the given data.

  ## Parameters

  * `data`: The data to make predictions on.

  ## Returns

  The predictions for the given data.
  """
  def predict(model, data) do
    for sequence in data do
      z = Weights * sequence + Biases
      predictions = tanh(z)
    end
    predictions
  end
end

File: rnn/zig/rnn.zig
struct RNN {
  input_dim: usize,
  hidden_dim: usize,
  output_dim: usize,

  weights: [usize; input_dim]<[usize; hidden_dim]f32],
  biases: [usize; hidden_dim]f32,

  h: [usize; hidden_dim]f32,
}

impl RNN {
  fn new(input_dim: usize, hidden_dim: usize, output_dim: usize) -> Self {
    Self {
      input_dim,
      hidden_dim,
      output_dim,

      weights: [
        [
          rand::random() - 0.5; hidden_dim
        ]
        for _ in 0..input_dim
      ],
      biases: [rand::random() - 0.5; hidden_dim],

      h: [0.0; hidden_dim],
    }
  }

  fn train(&mut self, data: &[[f32]]) {
    for sequence in data {
      for t in sequence {
        let mut z = 0.0;
        for i in 0..self.input_dim {
          z += t[i] * self.weights[i][self.h.len()];
        }
        self.h.push(tanh(z + self.biases[self.h.len()]));
      }
    }
  }

  fn predict(&mut self, data: &[f32]) -> [f32] {
    let mut predictions = [];
    self.h = [0.0; self.hidden_dim];
    for t in data {
      let mut z = 0.0;
      for i in 0..self.input_dim {
        z += t[i] * self.weights[i][self.h.len()];
      }
      self.h.push(tanh(z + self.biases[self.h.len()]));
      predictions.push(self.h[self.h.len() - 1]);
    }
    return predictions;
  }
}
